jerry-core/vm/vm.c

```diff
@@ -136,36 +136,55 @@ vm_op_set_value (ecma_value_t object, /**< base object */
                  ecma_value_t value, /**< ecma value */
                  bool is_strict) /**< strict mode */
 {
-  ecma_value_t completion_value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY);
-
-  ECMA_TRY_CATCH (obj_val,
-                  ecma_op_to_object (object),
-                  completion_value);
+  if (unlikely (!ecma_is_value_object (object)))
+  {
+    ecma_value_t to_object = ecma_op_to_object (object);
+    ecma_free_value (object);
 
-  ECMA_TRY_CATCH (property_val,
-                  ecma_op_to_string (property),
-                  completion_value);
+    if (ECMA_IS_VALUE_ERROR (to_object))
+    {
+      ecma_free_value (property);
+      return to_object;
+    }
 
-  ecma_object_t *object_p = ecma_get_object_from_value (obj_val);
-  ecma_string_t *property_p = ecma_get_string_from_value (property_val);
+    object = to_object;
+  }
 
-  if (ecma_is_lexical_environment (object_p))
+  if (!ecma_is_value_string (property))
   {
-    completion_value = ecma_op_put_value_lex_env_base (object_p,
-                                                       property_p,
-                                                       is_strict,
-                                                       value);
+    ecma_value_t to_string = ecma_op_to_string (property);
+    ecma_fast_free_value (property);
+
+    if (ECMA_IS_VALUE_ERROR (property))
+    {
+      ecma_free_value (object);
+      return to_string;
+    }
+
+    property = to_string;
   }
-  else
+
+  ecma_object_t *object_p = ecma_get_object_from_value (object);
+  ecma_string_t *property_p = ecma_get_string_from_value (property);
+  ecma_value_t completion_value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY);
+
+  if (!ecma_is_lexical_environment (object_p))
   {
     completion_value = ecma_op_object_put (object_p,
                                            property_p,
                                            value,
                                            is_strict);
   }
+  else
+  {
+    completion_value = ecma_op_set_mutable_binding (object_p,
+                                                    property_p,
+                                                    value,
+                                                    is_strict);
+  }
 
-  ECMA_FINALIZE (property_val);
-  ECMA_FINALIZE (obj_val);
+  ecma_free_value (object);
+  ecma_free_value (property);
```

Comment:
Could you please add a note saying that the `object` and `property` arguments will be freed by the function, and caller doesn't have to?